### DISCUSSION TITLE
Kyber example

### DISCUSSION THREAD

**Discussion by yanboti** [randombit/botan#3889] [Q&A] [answered]
Hello, I would like to understand why src/example/kyber.cpp uses "HKDF(SHA-512)" as the KDF. Should it not be SHAKE-256 as in the standard as specified in the submission?

**Comment by reneme** [comment]
There are two layers here: Internally, Kyber uses SHAKE for various things (including key derivation), and SHA-3 for ordinary hashing. The specified KDF in the example (`src/examples/kyber.cpp`) is applied to the shared secret that is output by Kyber. This _does not_ alter the Kyber algorithm itself. If you want to obtain the verbatim output of Kyber, pass "Raw" instead of "HKDF(SHA-512)" (and omit passing a salt to `encrypt()` and/or `decrypt()`).

Botan generally allows passing the output of any key agreement (or key encapsulation) mechanism through a KDF. Historically (e.g. for Diffie-Hellman) this was generally a good idea. For Kyber you don't have to do that anymore, but the library still allows you to do it for API consistency.

**Reply by yanboti** [reply]
Hello there, thank you for the reply.

`const std::string kdf = "Raw"; `

Did not work for me. May I know what exactly is the syntax?

**Reply by reneme** [reply, marked as answer]
Perhaps you need to remove the `salt` from the `encrypt()` and `decrypt()` calls? The following works for me:

```C++
#include <botan/kyber.h>
#include <botan/pubkey.h>
#include <botan/system_rng.h>
#include <array>
#include <iostream>

int main() {
   const size_t shared_key_len = 32;
   const std::string kdf = "Raw";

   Botan::System_RNG rng;

   Botan::Kyber_PrivateKey priv_key(rng, Botan::KyberMode::Kyber512_R3);
   auto pub_key = priv_key.public_key();

   Botan::PK_KEM_Encryptor enc(*pub_key, kdf);

   const auto kem_result = enc.encrypt(rng, shared_key_len);

   Botan::PK_KEM_Decryptor dec(priv_key, rng, kdf);

   auto dec_shared_key = dec.decrypt(kem_result.encapsulated_shared_key(), shared_key_len);

   if(dec_shared_key != kem_result.shared_key()) {
      std::cerr << "Shared keys differ\n";
      return 1;
   }

   return 0;
}
```

**Reply by yanboti** [reply]
Removing the salt works for me, but I would like to keep that too.

**Reply by reneme** [reply]
Not sure I understand. The salt is an input to the KDF (i.e. `HKDF(SHA-512)`). If you don't use a KDF (by passing "Raw"), you basically use vanilla Kyber and Kyber doesn't have the option to use a user-defined salt.

**Reply by yanboti** [reply]
I managed it by following the example in src/test/test_kyber.cpp and managed to get what I needed. Thank you for your help!
